l1infotreesync/processor.go

```
@@ -193,20 +190,33 @@ func (p *processor) GetLatestInfoUntilBlock(ctx context.Context, blockNum uint64
 		}
 	}()
 
-	lpb, err := p.getLastProcessedBlockWithTx(tx)
-	if err != nil {
-		return nil, err
+	if blockNum != nil {
+		if *blockNum == 0 {
+			return nil, ErrNoBlock0
+		}
+		lpb, err := p.getLastProcessedBlockWithTx(tx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve the last processed block: %w", err)
+		}
+		if lpb < *blockNum {
+			return nil, ErrBlockNotProcessed
```

Do we have to return an error here? The function description says it will return the most recent `L1InfoTreeLeaf` that occurred before or at `blockNum`. If block is not yet processed, shouldn't we return what is the latest we have?

---

The previous behaviour was returning the same error `ErrBlockNotProcessed`, if it's changed maybe we can have unexpected bug, WDYT?

---

I would probably leave it like this, because it was the behavior from before. Although the both ways have sense to some extent. However, probably this approach comes as a less surprise (IMO it is better to return an error, in case certain target block is not indexed than to return a wrong data).

---

To me it just doesn't make sense, because in the query, we also say `WHERE block_num <= $1` which means, it will get the latest it has until that block.